sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala

```diff
@@ -120,7 +120,7 @@ class GenericArrayData(val array: Array[Any]) extends ArrayData {
             if (!o2.isInstanceOf[Double] || ! java.lang.Double.isNaN(o2.asInstanceOf[Double])) {
               return false
             }
-          case _ => if (!o1.equals(o2)) {
+          case _ => if (o1.getClass != o2.getClass || o1 != o2) {
```

Comment:
ditto: Is this change above related to this PR?

Reply:
Yes. We cannot use `equals` and we cannot use `!=` here due to the following.
```scala
scala> Float.NaN == Float.NaN
val res2: Boolean = false

scala> Float.NaN.equals(Float.NaN)
val res3: Boolean = true
```